File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/aop/ControllerAspect.java
package com.windea.demo.cloudcollect.core.aop;

import com.windea.demo.cloudcollect.core.exception.ValidationException;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.springframework.validation.BindingResult;

/**
 * 控制器的切面。
 */
@Aspect
public class ControllerAspect {
	/**
	 * 参数验证的切面（在bindingResult有错时，抛出ValidationException）。
	 */
	@Before("within(com.windea.demo.cloudcollect.core.controller..*) && args(bindingResult, ..)")
	public void bindingResultAdvice(BindingResult bindingResult) {
		if(bindingResult.hasErrors()) {
			throw new ValidationException(bindingResult.getAllErrors());
		}
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/aop/ControllerAspect.kt
package com.windea.demo.cloudcollect.core.aop

import com.windea.demo.cloudcollect.core.exceptions.*
import org.aspectj.lang.*
import org.aspectj.lang.annotation.*
import org.springframework.stereotype.*
import org.springframework.validation.*

/**控制器的切面。*/
@Component
@Aspect
class ControllerAspect {
	/**参数验证的切面。在bindingResult有错时，抛出ValidationException。*/
	@Before("execution(* com.windea.demo.cloudcollect.core.controller..*(..,org.springframework.validation.BindingResult,..))")
	fun validationAdvice(joinPoint: JoinPoint) {
		//NOTE：当需要匹配任意位置的参数时，将类型写在execution表达式里面，而非将参数名写在args表达式里面
		joinPoint.args.filterIsInstance<BindingResult>().forEach { bindingResult ->
			if(bindingResult.hasErrors()) {
				throw ValidationException(bindingResult.allErrors)
			}
		}
	}
}
